sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/InputFileBuilder.java

```diff
@@ -93,19 +98,52 @@ DefaultInputFile create(File file) {
   DefaultInputFile completeAndComputeMetadata(DefaultInputFile inputFile, InputFile.Type type) {
     inputFile.setType(type);
     inputFile.setModuleBaseDir(fs.baseDir().toPath());
-    inputFile.setCharset(fs.encoding());
+    Charset charset = detectCharset(inputFile.file(), fs.encoding());
```

Comment:
Charset detection should be done after language detection. This would prevent opening every file, even if the file will be skipped.